packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -437,105 +477,100 @@ export class TokenDetectionController extends StaticIntervalPollingController<
+    const {
+      chainId: chainIdAgainstWhichToDetect,
+      networkClientId: networkClientIdAgainstWhichToDetect,
+    } = this.#getCorrectChainIdAndNetworkClientId(networkClientId);
+
+    if (!isTokenDetectionSupportedForNetwork(chainIdAgainstWhichToDetect)) {
       return;
     }
-    const selectedAddress = accountAddress ?? this.#selectedAddress;
-    const chainId = this.#getCorrectChainId(networkClientId);
 
     if (
       !this.#isDetectionEnabledFromPreferences &&
-      chainId !== ChainId.mainnet
+      chainIdAgainstWhichToDetect !== ChainId.mainnet
     ) {
       return;
     }
     const isTokenDetectionInactiveInMainnet =
-      !this.#isDetectionEnabledFromPreferences && chainId === ChainId.mainnet;
+      !this.#isDetectionEnabledFromPreferences &&
+      chainIdAgainstWhichToDetect === ChainId.mainnet;
     const { tokensChainsCache } = this.messagingSystem.call(
       'TokenListController:getState',
     );
-    const tokenList = tokensChainsCache[chainId]?.data ?? {};
-
+    const tokenList =
+      tokensChainsCache[chainIdAgainstWhichToDetect]?.data ?? {};
     const tokenListUsed = isTokenDetectionInactiveInMainnet
       ? STATIC_MAINNET_TOKEN_LIST
       : tokenList;
 
     const { allTokens, allDetectedTokens, allIgnoredTokens } =
       this.messagingSystem.call('TokensController:getState');
-    const tokens = allTokens[chainId]?.[selectedAddress] ?? [];
-    const detectedTokens = allDetectedTokens[chainId]?.[selectedAddress] ?? [];
-    const ignoredTokens = allIgnoredTokens[chainId]?.[selectedAddress] ?? [];
-
+    const [tokensAddresses, detectedTokensAddresses, ignoredTokensAddresses] = [
+      allTokens,
+      allDetectedTokens,
+      allIgnoredTokens,
+    ].map((tokens) =>
+      (
+        tokens[chainIdAgainstWhichToDetect]?.[addressAgainstWhichToDetect] ?? []
+      ).map((value) => (typeof value === 'string' ? value : value.address)),
+    );
     const tokensToDetect: string[] = [];
     for (const tokenAddress of Object.keys(tokenListUsed)) {
       if (
-        !findCaseInsensitiveMatch(
-          tokens.map(({ address }) => address),
-          tokenAddress,
-        ) &&
-        !findCaseInsensitiveMatch(
-          detectedTokens.map(({ address }) => address),
-          tokenAddress,
+        [
+          tokensAddresses,
+          detectedTokensAddresses,
+          ignoredTokensAddresses,
+        ].every(
+          (addresses) =>
+            !addresses.find((address) =>
+              isEqualCaseInsensitive(address, tokenAddress),
+            ),
         )
       ) {
         tokensToDetect.push(tokenAddress);
       }
     }
-    const sliceOfTokensToDetect = [];
-    sliceOfTokensToDetect[0] = tokensToDetect.slice(0, 1000);
-    sliceOfTokensToDetect[1] = tokensToDetect.slice(
+    const slicesOfTokensToDetect = [];
+    slicesOfTokensToDetect[0] = tokensToDetect.slice(0, 1000);
+    slicesOfTokensToDetect[1] = tokensToDetect.slice(
```

Comment:
I've been wondering why its bucketed into 2 buckets this way.  On mainnet for example, it gets bucketed into slices of 1,000 and 3,000:

<img width="340" alt="Screenshot 2024-02-22 at 1 25 43 PM" src="https://github.com/MetaMask/core/assets/3500406/f4ad9f21-3cb0-4182-9941-85e8cb6f735e">

But I'm not sure why the bucketing was originally there, to determine whether it should be four even buckets of 1000.

Reply:
Oh that's not good. It should be four even buckets. I've heard that this section was written the way it is because originally 2000 was the max expected number of tokens.

We have a ticket for fixing this: https://github.com/MetaMask/core/issues/3661. I'll suggest bumping its priority.